src/BotPlutusInterface/Balance.hs

```diff
@@ -65,71 +69,127 @@ import Plutus.V1.Ledger.Api (
  )
 
 import BotPlutusInterface.BodyBuilder qualified as BodyBuilder
-import Data.Bifunctor (bimap)
+
+-- import BotPlutusInterface.CollateralEff qualified as CollateralEff
+
+import BotPlutusInterface.Collateral (removeCollateralFromMap)
 import Prettyprinter (pretty, viaShow, (<+>))
 import Prelude
 
+-- `BalanceTx` is for checking which type of `Tx` we are balancing.
+data BalanceTx
+  = -- | `BalanceTxCollateral` denotes that we are balancing the `Tx` that creates collateral.
+    BalanceTxCollateral
+  | -- | `BalanceTxWithScripts` denotes that we are balancing the `Tx` that uses scripts.
+    BalanceTxWithScripts
```

Comment:
Is it worth the added complexity to maintain a distinction between transactions that contain scripts or not?

Reply:
If we don't distinguish which kind of `Tx` we are balancing early, we will have to check use these checks like `validCollateralTx` in `BalanceTxIO'` and then case split on conditions anyway, this leads lot of checking in the `balanceTxIO'` and `utxosAndCollateralAtAddress`.